doc/l2mc/IGMP_Snooping_HLD.md

```diff
@@ -0,0 +1,701 @@
+
+
+## 1.1 Requirements
+
+### 1.1.1 Functional Requirements
+ 1. Support IGMP Snooping for IGMP Version1 (IGMPv1), IGMPv2, IGMPv3 (INCL Mode).
+ 2. Support IPv4 based forwarding.
+ 3. Support IGMP fast leave.
+ 4. Support IGMP querier functionality.
+ 5. Support static and dynamic multicast router list. 
+ 6. Aging of dynamic mrouter ports will be supported. Static mrouter ports will not timed out, they have to be un-configured explicitly.
+ 7. Support static multicast group configuration. Static l2mc entries will not timed out, they have to be un-configured explicitly.
+ 8. Support trapping IGMP and PIM Hello control packets to CPU.
+ 9. Support Link-layer topology changes (due to STP) and generate queries on all non-mrouter ports for faster convergence. 
+ 10. Support IS-CLI Configuration and show commands
+ 11. Support IGMP Snooping for MCLAG
+ 
+
+### 1.1.2 Configuration and Management Requirements
+
+To support IS-CLI configuration and show commands.
+ 1. Support CLI configurations as described in "Configuration Commands" section below
+ 2. Support show commands as described in "Show Commands" section below
+ 3. Support openconfig REST API and gNMI
+
+### 1.1.3 Scalability Requirements
+
+SONiC IGMP Snooping feature does not have scale limit, scale number varies based on ASIC supported number, refer to scalability section below for the tested scale numbers.
+ 
+### 1.1.4 Warm Boot Requirements
+
+Dynamic and static L2MC entries should persist across warm reboot with no traffic disruption to the active flows.
+And, remote learnt L2MC entries in MCLAG should persist across warm reboot.
+Learning of new snooping entries and aging of existing L2MC entries will stop while the control plane is away.
+- To support planned system warm boot.
+
+## 1.2 Design Overview
+
+### 1.2.1 Basic Approach
+### 1.2.2  L2MC Docker
+A new L2MC docker will be created to support IGMP snooping.
+L2MC will host l2mcastMgr and l2mcastd,  and l2mcastSync will be part of l2mcastd, the details of the changes will be discussed in the design section below.  
+
+### 1.2.3 IGMP control packets flooding 
+IGMP control packets flooding happens in kernel, kernel  flooding behaviour is per bridge, there is no per VLAN flooding control in Kernel.  Until the dynamic multicast-router(mrouter) port learns in the VLAN, IGMP reports will be flooded to all members of the VLAN.  
+In presence of multicast-router, IGMP reports received on a host interface are forwarded to multicast-router interfaces alone, but not to the other host interfaces in the VLAN.
+
+### 1.2.4 SAI Overview
```

Comment:
Could you please provide more details on how the SAI apis are going to be invoked? Does the sync layer maintain seperate database of L2MC entries and their port member lists? How does it going to differentiate new L2MC entry creation versus adding a port member to existing group? It would be good to put an example how the sync layer interacts with SAI layer wrt programming of the entries?

Reply:
Yes, L2mcorch in Orchagent maitains database of L2MC entries and their port member information. L2MC (Source,Group,Vlan) entry will be created in L2mcorch when it recieves first member addition, subsequent member add will bring in source group and the vlan information, based on that update to an already existing entry will be decided. L2mcorch will call sairedis API's to program L2mc entry, L2mc group and L2mc group members.

Example: 
APP_L2MC_MEMBER_TABLE:Vlan10:0.0.0.0:225.0.0.1:Port1
L2mcorch processes and creates db entry when L2MC entry add comes to APPDB for the first time on the port1, key for db lookup here will be Source, Group and the Vlan. 
L2mcorch will call sairedis API to create L2MC group first, then it calls API to create L2MC entry in the hardware and later L2MC group member add API will be called to update new port list.

When there is new receiver joins on Port1 on the same Vlan10 as mentioned below, lookup will find an already existing L2MC entry, this will update new port2 to the mc_id by calling sairedis API to add new group member.
APP_L2MC_MEMBER_TABLE:Vlan10:0.0.0.0:225.0.0.1:Port2

I will put these details.